mlx/fast.cpp

```diff
@@ -559,15 +562,25 @@ array scaled_dot_product_attention(
   };
 
   auto stream = to_stream(s);
-  constexpr const int supported_head_dim = 128;
   const size_t query_head_dim = q.shape(-1);
+  const bool supported_head_dim =
+      query_head_dim == 64 || query_head_dim == 80 || query_head_dim == 128;
+
+  const bool supported_head_dim_self_attn =
+      query_head_dim == 64 || query_head_dim == 128;
   const size_t query_sequence_length = q.shape(2);
-  bool implementation_supports_use_case = batch_dim == 1 &&
-      query_sequence_length == 1 && !mask.has_value() &&
-      query_head_dim == supported_head_dim && final_type != bfloat16 &&
+  const bool full_self_attention = query_sequence_length >= 16 &&
+      !mask.has_value() && supported_head_dim_self_attn &&
+      n_q_heads == n_kv_heads && final_type != bfloat16 &&
       stream.device == Device::gpu;
-  // TODO, update routing conditions post further tuning
+
+  // fast decoding gpu shader
+  bool implementation_supports_use_case = batch_dim == 1 &&
+      query_sequence_length == 1 && !mask.has_value() && supported_head_dim &&
+      final_type != bfloat16 && stream.device == Device::gpu;
   implementation_supports_use_case &= false;
```

Comment:
Could you change the condition names slightly. Something like:

- `supports_sdpa`
- `supports_full_self_attention`

Then `implementation_supports_use_case = supports_sdpa | supports_full_self_attention`.

I think it will be a bit more clear what's going on there.